[Merged by Bors] - feat(analysis/normed_space/inner_product): Bessel's inequality

gihanmarasingha opened the pull request:
A proof both of Bessel's inequality and that the infinite sum defined by Bessel's inequality converges.



---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

gihanmarasingha committed:
feat(analysis/normed_space/inner_product): Bessel's inequality
A proof both of Bessel's inequality and that the infinite sum defined by Bessel's inequality converges.

gihanmarasingha committed:
remove unused instances of `inner_product space 𝕜 E`, `is_R_or_C 𝕜` and `complete_space E`.

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,56 @@ linear_map.mk_continuous
+
+variables {ι: Type*} (x : E) {v : ι → E}
+
+lemma bessel_finite {s : finset ι} (hv : orthonormal 𝕜 v) : ∑ i in s, ∥ ⟪v i, x⟫ ∥^2 ≤ ∥x∥ ^2 :=
```
The convention is to make the name descriptive of the contents, and then put the "human" name in a docstring above.  Here, perhaps `orthonormal.sum_inner_products_le` would work (and `orthonormal.tsum_inner_products_le`, `orthonormal.inner_products_summable` for the others) -- putting them in the namespace `orthonormal` allows for the use of projection notation with the variable `hv`, as we were discussing yesterday.

hrmacbeth replied:
```suggestion
lemma bessel_finite {s : finset ι} (hv : orthonormal 𝕜 v) : ∑ i in s, ∥⟪v i, x⟫∥ ^ 2 ≤ ∥x∥ ^ 2 :=
```
Spacing, here and below ... the convention is that `∥` surrounds things without spaces, `^` gets spaces.

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,56 @@ linear_map.mk_continuous
+  simp only [inner_smul_left, inner_smul_right, inner_conj_sym, ←mul_assoc],
+  have h₂ : ∑ i in s, ∑ i_1 in s, (inner (v i) x * inner x (v i_1)) * inner (v i_1) (v i)
+   = (∑ k in s, (inner (v k) x * inner x (v k)) : 𝕜 ),
+   { apply orthonormal.inner_left_right_finset, exact hv },
```
You can use the projection notation here!
```suggestion
   { exact hv.inner_left_right_finset },
```

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,56 @@ linear_map.mk_continuous
+  simp only [two_mul, add_sub_cancel'],
+  have h₃ : ∀ z : 𝕜, re (z * conj (z)) = ∥ z ∥ ^2,
+  { intro z, simp only [mul_conj, norm_sq_eq_def'], norm_cast, },
+  simp only [←h₃, inner_conj_sym],
```
In this proof, if you move `h₂` and `h₃` upwards, I think you'll find that almost all of the last series of `simp only`s can be combined into a single `simp`.

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,56 @@ linear_map.mk_continuous
+begin
+  rw ←sub_nonneg,
+  suffices hbf: ∥ x -  ∑ i in s, ⟪v i, x⟫ • (v i) ∥^2 = ∥x∥^2 - ∑ i in s, ∥ ⟪v i, x⟫ ∥^2,
+  { rw ←hbf, simp only [norm_nonneg, pow_nonneg], },
```
With some exceptions, the style is to put each tactic application on a new line -- this applies both here and in a few places below.

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,56 @@ linear_map.mk_continuous
+begin
+  refine tsum_le_of_sum_le' _ _,
+  { simp only [norm_nonneg, pow_nonneg], },
+  { intro s, apply bessel_finite _ hv }
```
You'll be able to use the projection notation once you implement the name change I suggested, so the whole proof can become something like
```suggestion
  refine tsum_le_of_sum_le' _ (hv.sum_inner_products_le x),
  simp only [norm_nonneg, pow_nonneg]
```

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,56 @@ linear_map.mk_continuous
+    { apply is_lub_cSup,
+      { apply set.range_nonempty, },
+      { refine @set.nonempty_of_mem _ _ (∥x∥^2) _,
+        rw mem_upper_bounds,
```
It turns out that both `set.nonempty_of_mem` and `mem_upper_bounds` are lemmas which rewrite down to the definitions, so you can actually just skip their application and use the definitions directly.
```suggestion
      { use ∥x∥ ^ 2,
```

hrmacbeth reviewed (commented):
Thanks for this!  A bunch of style comments, everyone has these on their first PRs.

gihanmarasingha committed:
Apply suggestions from code review

Style changes regarding use of projection notation and use of space.

Co-authored-by: hrmacbeth <25316162+hrmacbeth@users.noreply.github.com>

gihanmarasingha committed:
style changes.
Add doc strings.
Change function names to be more descriptive and to use projection notation.

gihanmarasingha committed:
refactor `orthonormal.sum_inner_products_le`
Proof proof of `h₂` and `h₃` up. Combine multiple
`simp only` applications.
Fix spacing. Put each tactic on a line by itself.

gihanmarasingha committed:
refactor `orthonormal.tsum_inner_products_le`
Use projection notation to shorten proof of Bessel's inequality.

gihanmarasingha committed:
style in `orthonormal.inner_products_summable`
Put each tactic application on a line by itself.

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,61 @@ linear_map.mk_continuous
+  { rw ←hbf,
+    simp only [norm_nonneg, pow_nonneg], },
+  rw [norm_sub_sq, sub_add],
+  congr' 1,
```
I think the rest of the proof works without this `congr' 1` line.

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,61 @@ linear_map.mk_continuous
+
+/-- Bessel's inequality. -/
+lemma orthonormal.tsum_inner_products_le (hv : orthonormal 𝕜 v)
+  : ∑' i, ∥⟪v i, x⟫∥ ^ 2 ≤ ∥x∥ ^ 2 :=
```
When breaking a lemma statement between hypotheses and goal, mathlib style is to put the `:` on the previous line.

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,61 @@ linear_map.mk_continuous
+    { intro b,
+      simp only [norm_nonneg, pow_nonneg], },
+    { apply is_lub_cSup,
+      { apply set.range_nonempty, },
```
Personally I would merge these two:
```suggestion
    { refine is_lub_cSup (set.range_nonempty _) _,
```
(with some rearrangement of the braces on the following lines in consequence).  But your way is also perfectly good style, so do as you please.

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,61 @@ linear_map.mk_continuous
+      { apply set.range_nonempty, },
+      { use ∥x∥ ^ 2,
+        rintro y ⟨s, rfl⟩,
+        apply orthonormal.sum_inner_products_le _ hv, }, }, },
```
```suggestion
        exact hv.sum_inner_products_le x }, }, },
```

hrmacbeth reviewed (commented):
This is looking good!  Just a couple more points.

gihanmarasingha committed:
remove `congr' 1` in `orthonormal.sum_inner_products_le`

gihanmarasingha committed:
move colon to previous line

gihanmarasingha committed:
change two applications of `apply` to one application of `refine` in `orthonormal.inner_products_summable`

gihanmarasingha committed:
use projection notation in the proof of `orthonormal.inner_products_summable`

hrmacbeth commented on src/analysis/normed_space/inner_product.lean:
```diff
@@ -1374,6 +1382,61 @@ linear_map.mk_continuous
+    norm_cast, },
+  rw ←sub_nonneg,
+  suffices hbf: ∥x -  ∑ i in s, ⟪v i, x⟫ • (v i)∥ ^ 2 = ∥x∥ ^ 2 - ∑ i in s, ∥⟪v i, x⟫∥ ^ 2,
+  { rw ←hbf,
```
Did you miss this one, or prefer not to implement it?

gihanmarasingha replied:
I missed this before, but I'm not actually sure what's being suggested. Sorry! Would you kindly explain again? Thanks.

hrmacbeth replied:
Sorry for the ambiguity.  This is an opportunity to reduce 3 lines to 2 (which always makes us happy!). In this case, you can move the `rw ←sub_nonneg` inside the `suffices` block and the proof goes through without change -- and then, since the `suffices` block  would begin 
```lean
  { rw ←sub_nonneg,
    rw ←hbf,
```
we might as well combine these two lines, to
```lean
 { rw [←sub_nonneg, ←hbf],
```

gihanmarasingha committed:
combine two `rw`s
